In [2]:
import numpy as np

**1. Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8**

In [3]:
rand_array = np.array([5, -2, 7, 0, 8, -6, 3, 9, -1, 4])
sign_change=np.where((rand_array>=3) & (rand_array<=8), rand_array*(-1), rand_array)
print(sign_change)

[-5 -2 -7  0 -8 -6 -3  9 -1 -4]


**2. Заменить максимальный элемент случайного массива на 0**

In [ ]:
rand_array[rand_array.argmax()]=0
print(rand_array)

[ 5 -2  7  0  8 -6  3  0 -1  4]


**🫵🏼3. Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив**

In [ ]:
arr1=np.random.randint(-20, 20, size=(2, 2))
arr2=np.random.randint(-30, 20, size=(2, 2))

a = arr1.ravel()
b = arr2.ravel()

mul=np.stack(np.meshgrid(arr1, arr2), axis=-1).reshape(-1, 2)
print('Array 1:', arr1)
print('Array 2:', arr2)
print('Result: ', mul)

Array 1: [ -3  14 -13   2]
Array 2: [-30 -21 -26 -23]
Result:  [[ -3 -30]
 [ 14 -30]
 [-13 -30]
 [  2 -30]
 [ -3 -21]
 [ 14 -21]
 [-13 -21]
 [  2 -21]
 [ -3 -26]
 [ 14 -26]
 [-13 -26]
 [  2 -26]
 [ -3 -23]
 [ 14 -23]
 [-13 -23]
 [  2 -23]]


**4. Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B**

In [ ]:
A = np.random.randint(-3, 3, size=(8, 3))
B = np.random.randint(-2, 2, size=(2, 2))

matching_rows = []

for a, row_a in enumerate(A):
  for b, row_b in enumerate(B):
    if all(set(row_b).issubset(set(row_a)) for row_b in B):
      matching_rows.append(a)
      break

print("A =\n", A)
print("A =\n", B)
print(matching_rows)

A =
 [[-3  1 -3]
 [ 0 -1  1]
 [-2 -2 -1]
 [-1 -2 -3]
 [-2 -1  0]
 [-2  1 -1]
 [ 0 -3 -3]
 [ 0 -1 -2]]
A =
 [[1 1]
 [0 1]]
[1]


**5. Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)**

In [ ]:
array= np.array([[ 1,  2,  3],
                [ 4,  5,  6],
                [ 7,  8,  9],
                [1, 1, 1],
                [ 3,  2,  1],
                [ 5,  4,  6],
                [ 2, 2, 2],
                [ 1,  3,  2],
                [ 1,  3,  2],
                [ 1,  3,  2]])
# 1 way
for a, row_a in enumerate(array):
  if len(set(row_a))==1:
    modif_arr = np.delete(array, a, axis=0)

# 2 way (best)
mask = ~(array==array[:, [0]]).all(axis=1)
array[mask]

array([[1, 2, 3],
       [4, 5, 6],
       [7, 8, 9],
       [3, 2, 1],
       [5, 4, 6],
       [1, 3, 2]])

**6. Дан двумерный массив. Удалить те строки, которые повторяются**

In [ ]:
array= np.array([[ 1,  2,  3],
                [ 4,  5,  6],
                [ 7,  8,  9],
                [ 5,  4,  6],
                [1, 1, 1],
                [ 3,  2,  1],
                [ 5,  4,  6],
                [ 5,  4,  6],
                [ 2, 2, 2],
                [2, 2, 2],
                [1,  3,  2],
                [1,  3,  2]])
# 1 way
mask = np.unique(array, axis=0, return_index=True)[1]
array[mask]

# 2 way
array=np.unique(array, axis=0)

array([[1, 1, 1],
       [1, 2, 3],
       [1, 3, 2],
       [2, 2, 2],
       [3, 2, 1],
       [4, 5, 6],
       [5, 4, 6],
       [7, 8, 9]])



---
---

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).

Замечание 1. Можно считать, что все указанные объекты непустые (к примеру, в задаче 1 на диагонали матрицы есть ненулевые элементы).

Замечание 2. Для большинства задач решение занимает не больше 1-2 строк.

---

---




Задача 1: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.
Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
X = np.array([[1, 0, 1],
              [2, 0, 2],
              [3, 0, 3],
              [4, 4, 4]])

# numpy
result = np.prod(main_diagonal[np.diag(X)!=0])
print('Numpy result ', result)

# python
result2=1
for i in range(min(len(X), len(X[0]))):
  if X[i][i]!=0:
    result2 *= X[i][i]

print('Python result ', result2)

Numpy result  3
Python result  3


Задача 2: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.
Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [ ]:
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])

# numpy
np_check = np.equal(np.sort(x), np.sort(y)).all()

# python
sorted(x)==sorted(y)

True

Задача 3: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль. Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [ ]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

# numpy
# after_null = (np.where(x==0)[0]+1)<len(x)
# after_null = after_null[after_null<len(x)]

# max_elem = np.max(x[after_null])

# python
after_null=[]
for i in range(len(x)):
  if x[i]==0:
    after_null.append(i+1)

after_null = [i for i in after_null if i<len(x)]
elements = [x[i] for i in after_null]
max_elem = max(elements)
max_elem

np.int64(5)

Задача 4: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.
Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [19]:

x = np.array([2, 2, 2, 3, 3, 3, 5])

# numpy
unique_values, count = np.unique(x, return_counts=True)
print(f'Unique array: {unique_values}\n Count: {count}')

# python
value = []
counts = []

current_value=x[0]
current_count=1

for i in range(1, len(x)):
  if x[i]==current_value:
    current_count +=1
  else:
    value.append(current_value)
    counts.append(current_count)
    current_value = x[i]
    current_count =1

value.append(current_value)
counts.append(current_count)

print(value, counts)

[np.int64(2), np.int64(3), np.int64(5)] [3, 3, 1]


Задача 5: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [29]:
import time
from scipy.spatial.distance import cdist

X = np.array([
    [1, 2],
    [4, 6],
    [7, 8]
])

Y = np.array([
    [2, 5],
    [7, 9],
    [0, 17]
])

# numpy
start_time = time.time()
D=np.zeros((len(X), len(Y)))

for i in range(len(X)):
  for j in range(len(Y)):
    D[i, j]=np.linalg.norm(X[i]-Y[j])

print(f'D = \n{D}\n')
print('Time: ', time.time()-start_time)

# scipy
start = time.time()

D2 = cdist(X, Y, metric='euclidean')
print(f'\nD2 = \n{D2}\n')
print("Time scipy:", time.time() - start)

D = 
[[ 3.16227766  9.21954446 15.03329638]
 [ 2.23606798  4.24264069 11.70469991]
 [ 5.83095189  1.         11.40175425]]

Time:  0.003755331039428711

D2 = 
[[ 3.16227766  9.21954446 15.03329638]
 [ 2.23606798  4.24264069 11.70469991]
 [ 5.83095189  1.         11.40175425]]

Time scipy: 0.0006749629974365234


**Задача 6: CrunchieMunchies **

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием CrunchieMunchies.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши СrunchieMunchies - самый здоровый выбор для потребителей.

In [31]:
calorie_stats = np.loadtxt("./cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?
Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [40]:
average_calories = (np.mean(calorie_stats)-60).round(2)
print(average_calories)

46.88


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.
Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [38]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.
Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [42]:
median_calories = np.median(calorie_stats)
print(median_calories)

110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.
Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [55]:
percentiles = np.arange(0, 101, 1)

values=np.percentile(calorie_stats, percentiles)

for i, j in zip(percentiles, values):
  if j>60:
    nth_percentile=i
    break

print(f'Min percentile more then 60 calories: {nth_percentile}')

Min percentile more then 60 calories: 4


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.
Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [68]:
more_calories = ((calorie_stats>60).mean()*100).round(2)
print(f'More then 60 calories by {more_calories}% of competitors')

More then 60 calories by 96.1% of competitors


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?
Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [71]:
calorie_std = np.std(calorie_stats).round(2)
print(calorie_std)

19.36


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Полученные данные показывают, что CrunchieMunchies имеет явное преимущество перед злаками конкурентов по количеству калорий. Средняя калорийность конкурентов примерно на 47 калорий выше, что составляет почти 80% по сравнению с одной порцией CrunchieMunchies. Кроме того, около 96% конкурирующих продуктов содержат больше калорий. Результаты исследования могут быть использованы в маркетинговых кампаниях, рекламных слоганах и промоматериалах для наглядной демонстрации реальных отличий продукта от конкурентов.

